File: src/svg_painter.cpp
#include "svg_painter.h"

#include <assert.h>

using namespace std;

SvgPainter::SvgPainter(ostream& os, const PainterParams& params): os_(os) {
    char buf[128];
    snprintf(buf, sizeof(buf), "stroke:black;stroke-width:%d;stroke-linecap:round", params.stroke_width);
    wallStyle_ = buf;
    snprintf(buf, sizeof(buf), "stroke:red;stroke-width:%d;stroke-linecap:round", params.stroke_width);
    wallBlockedStyle_ = buf;
}

void SvgPainter::BeginDraw(int width, int height) {
    os_ << "<!DOCTYPE svg>\n";
    os_ << "<svg height=\"" << height << "\" width=\"" << width << "\" xmlns=\"http://www.w3.org/2000/svg\" style=\"background-color:white\">\n";
}

void SvgPainter::EndDraw() {
    os_ << "</svg>\n";
}

string SvgPainter::toSvgStyle(IPainter::EStyle style) const {
    switch (style) {
        case IPainter::EStyle::OpenCell: return "fill:darkgray";
        case IPainter::EStyle::VisitedCell: return "fill:white";
        case IPainter::EStyle::OnPathCell: return "fill:lightgray";
        case IPainter::EStyle::Wall: return wallStyle_;
        case IPainter::EStyle::WallBlocked: return wallBlockedStyle_;
        default:
            assert(0);
    }
}

void SvgPainter::DrawLine(const Point2D& p1, const Point2D& p2, EStyle style) {
    os_ << "<line x1=\"" << p1.x
        << "\" y1=\"" << p1.y
        << "\" x2=\"" << p2.x
        << "\" y2=\"" << p2.y
        << "\" style=\"" << toSvgStyle(style) << "\" />\n";
}

void SvgPainter::DrawPoly(const vector<Point2D>& vertices, EStyle style) {
    os_ << "<polygon points=\"";
    for (const auto v : vertices) {
        os_ << v.x << ',' << v.y << ' ';
    }
    os_ << "\" style=\"" << toSvgStyle(style) << "\" />\n";
}

File: maze4kids/src/lib/maze/SvgPainter.ts
import { type Painter, PaintStyle, type Point2D } from './Painter';

interface PainterParams {
    strokeWidth: number;
}

const SVG_NAMESPACE = 'http://www.w3.org/2000/svg';

export class SvgPainter implements Painter {
    private node: Node;
    private params: PainterParams;
    private svg: SVGSVGElement | undefined;

    constructor(node: Node, params: PainterParams) {
        this.node = node;
        this.params = params;
    }

    BeginDraw(width: number, height: number): void {
        const svg = document.createElementNS(SVG_NAMESPACE, 'svg');
        svg.setAttribute('width', width.toString());
        svg.setAttribute('height', height.toString());
        svg.style.backgroundColor = 'white';
        this.svg = svg;
    }

    EndDraw(): void {
        this.node.appendChild(this.svg!);
    }

    DrawLine(p1: Point2D, p2: Point2D, style: PaintStyle): void {
        const line = document.createElementNS(SVG_NAMESPACE, 'line');
        line.setAttribute('x1', p1.x.toString());
        line.setAttribute('y1', p1.y.toString());
        line.setAttribute('x2', p2.x.toString());
        line.setAttribute('y2', p2.y.toString());
        this.setStyle(line.style, style);
        this.svg!.appendChild(line);
    }

    DrawPoly(vertices: Point2D[], style: PaintStyle): void {
        const polygon = document.createElementNS(SVG_NAMESPACE, 'polygon');
        polygon.setAttribute('points', vertices.map(v => `${v.x},${v.y}`).join(' '));
        this.setStyle(polygon.style, style);
        this.svg!.appendChild(polygon);
    }

    private setStyle(cssStyle: CSSStyleDeclaration, style: PaintStyle) {
      switch (style) {
        case PaintStyle.OpenCell:
          cssStyle.fill = 'white';
          break;

        case PaintStyle.VisitedCell:
          cssStyle.fill = 'white';
          break;

        case PaintStyle.OnPathCell:
          cssStyle.fill = 'lightgray';
          break;

        case PaintStyle.Wall:
          cssStyle.stroke = 'black';
          cssStyle.strokeWidth = this.params.strokeWidth.toString();
          cssStyle.strokeLinecap = 'round';
          break;

        case PaintStyle.WallBlocked:
          cssStyle.stroke = 'black';
          cssStyle.strokeWidth = this.params.strokeWidth.toString();
          cssStyle.strokeLinecap = 'round';
          break;
      }
    }
}
